Code hacker for unit tests

Konamiman opened the pull request:
## What is this?

This pull request provides a mechanism to ease unit testing of classes with language constructs that are hard or impossible to mock, such as:

* Calls to static methods in other classes.
* Use of standalone PHP, WP or Woo functions.
* `final` classes.

Please see [the README file](https://github.com/woocommerce/woocommerce/blob/32dac2486f3198471f7e108eaf7b209ceac46896/tests/Tools/CodeHacking/README.md) for the full details.

### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

Konamiman committed:
Barebones implementation of a code hacker for unit tests.

The "code hacker" is a class that hooks on filesystem events
(using stream_wrapper_unregister) in order to allow for dynamically
modifying the content of PHP code files while they are loaded.
The code hacker class allows registering hacks, which are
functions that take source code as input and return the modified code.
A hack can be a standalone function or a class with a "hack" method.

A few hacks are provided off the shelf. One allows mocking standalone
PHP functions (WP, WOO or not), another one allows mocking static
methods, and there's the one that removes the "final" qualifier
from a class definition. This helps unit testing stuff that would
otherwise be quite hard to test.

Konamiman committed:
Add the @hack code annotation for tests.

Now @hack class and method annotations can be used to register
code hacks as an alternative to using before_ methods.
The syntax is /* @hack HackClassName param1 param2 */
where parameters will be passed to the class constructor.
If the class name ends with "Hack", then that suffix can be
omitted (e.g. "Foo" can be specified instead of "FooHack").

Konamiman committed:
Miscellaneous code hacking fixes:

- Fix how CodeHackerTestHook::executeBeforeTest parses the test name,
  to account for warnings and tests with data sets.

- CodeHackerTestHook now includes a executeAfterTest hook that
  disables the code hacker (needed to prevent it from inadvertently
  altering further tests). Also, clear_hacks is executed in
  executeBeforeTest for the same reason.

- CodeHacker gets restore, clear_hacks and is_enabled methods
  to support the changes in CodeHackerTestHook.

- FunctionsMockerHack fixed so that it doesn't modify strings
  that are class method definitions.

- Added the WC_Unit_Test_Case::file_copy method, it must be used
  instead of the PHP built-in "copy" in tests, otherwise tests
  that run with the code hacker active will fail.
  This is something to investigate.

Konamiman committed:
All code hacking files moved to src\Testing folder.

Konamiman committed:
Fix code sniffer errors in CodeHacker and related classes.

Konamiman committed:
Add a README file for the code hacker.

Konamiman committed:
Add a workaround for code hacking static methods on already loaded files.

The unit testing bootstrap loads and initializes WooCommerce, this
loads a bunch of code files that can't then be hacked in the test hooks.

A workaround is provided in this commit for the case of hacking
static methods. A new StaticWrapper class is created that allows
defining mock methods after the code file has been loaded.
This is applied to all classes from a fixed list in the bootstrap,
before WooCommerce is initialized. The list should be kept up to date
with the list of classes that require such workaround.

vedanshujain reviewed (commented):
I took a cursory look and liked the idea in general. I think this will be really helpful in increasing our unit test library and we should aim to merge this PR sooner than later.
Some feedback/questions:
1. There are some code standard violation in the code. Perhaps new files are not getting checked by linters? We should fix this before the merge.
2. Why not include this library in the unit test suite bootstrap files, instead of having to include them when needed?
3. Can you document what are the limitations, i.e. what all cannot be mocked. Can this also mock global functions, PHP native functions etc? Can we mock `copy` method to make it work for all tests, if not, can we add a custom error to `copy` method somehow to indicate that our method should be instead?
4. Looks like tests are failing on Travis, we need to fix before we can merge.

Konamiman commented:
Hey @vedanshujain, thanks for your comments.

> There are some code standard violation in the code. Perhaps new files are not getting checked by linters? We should fix this before the merge.

That's weird, I don't get any error when running the code sniffer locally. What are the violations that you are finding?

> Why not include this library in the unit test suite bootstrap files, instead of having to include them when needed?

I'm not sure if I get what you mean here. You don't "include" the code hacker, rather you activate it when needed; activating means registering the hacker as a stream wrapper for the filesystem. Since this is a pretty bold thing to do (it's a "hack" after all) I wanted to activate it only when needed.

However after trying to use the hacker for real life tests I found an additional problem: some files are loaded during the bootstrap execution, and are therefore not hackable during the test hooks execution. This forced me introduce a workaround (see last commit) and effectively activate the hacker in the bootstrap.

> Can you document what are the limitations, i.e. what all cannot be mocked. Can this also mock global functions, PHP native functions etc? 

The code hacker is a generic mechanism that allows to alter PHP code. Hacks are provided for mocking static functions and global functions for convenience, but any new hack can be introduced if necessary. 
It's a last resort mechanism that should be used only when other proper mechanism (PHPUnit mocking, Woo helpers) don't work; the README file explains all of this (and if it's not clear enough, please tell me!)

So I think that the current approach in the README ("This allows you to hack anything, use responsibly, see the built-in hacks sources for details on how to use them") is the appropriate one, but of course the exact wording could be improved if necessary.

> Can we mock copy method to make it work for all tests, if not, can we add a custom error to copy method somehow to indicate that our method should be instead?

The code hacker mocks the tested code, not the tests themselves 🙂 (by the time the code hook runs the test file is already loaded so it's not hackable). 

The issue with `copy` is a very weird one that should be investigated, however since only a handful of tests are using it I thought it was enough to create the helper method as a workaround and to mention it in the documentation. Anyway, how could we add an error whenever it's used?

> Looks like tests are failing on Travis, we need to fix before we can merge.

Everything is failing in Travis with `Class 'Automattic\WooCommerce\Testing\CodeHacking\StaticWrapper' not found in /home/travis/build/woocommerce/woocommerce/tests/bootstrap.php:40` (StaticWrapper is a class that I added in the last commit), not sure why as it's working locally, perhaps the autoloader hasn't been run yet at that point?

claudiosanches commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,320 @@
+ * @package WooCommerce/Testing
+ */
+
+//phpcs:disable Squiz.Commenting.FunctionComment.Missing, Squiz.Commenting.VariableComment.Missing
```
This one helps with documentation, there's no reason to avoid including docblocks in functions, no code in WooCommerce is acceptable without proper inline docs.

Konamiman replied:
I hadn't included comments for this class because it's a stream wrapper class where most of the methods are irrelevant for the code hacking and are just transparent wrappers for the equivalent PHP filesystem functions. But if everything must have docs, then sure, I'll add them.

Konamiman replied:
...aaand done!

Konamiman committed:
Add function docs to the CodeHacker class.

Konamiman committed:
Improve error messaging in WC_Tests_MaxMind_Database::test_download_database_works

Konamiman committed:
Improvements on the code hacker.

- Add methods to temporarily disable and reenable the code hacker.

The code hacker is causing issues in some tests that perform
write operations to the local filesystem. Since this happens only
in a few cases, the easiest fix is to temporarily disable the
code hacker when that happens. This commit adds two new methods
for that in `WC_Unit_Test_Case`: `disable_code_hacker` and
`reenable_code_hacker`.

These methods use a disabling requests count so that the hacker
isn't enabled before it should. E.g. you call `disable`, then
a helper method that does `disable` and `enable`, then `enable` -
then only the last `enable` will have effect.

- `CodeHacker::add_hack` has now a boolean `persistent` parameter.
Persistent hacks won't be cleared by `clear_hacks`.

- `CodeHackerTestHook::executeAfterTest` will now disable the hacker
only if no persistent hacks are registered.

- The existing `file_copy` method is made static for consistency.

- `CodeHacker::restore` method renamed to `disable` for clarity.

vedanshujain commented on src/Testing/CodeHacking/CodeHackerTestHook.php:
```diff
@@ -0,0 +1,167 @@
+ *    use Automattic\WooCommerce\Testing\CodeHacking\CodeHacker;
+ *    use Automattic\WooCommerce\Testing\CodeHacking\Hacks\...
+ *
+ *    public static function before_all($method_name) {
```
Why don't we use `setup()` instead of having to implement `before_all`?

vedanshujain replied:
I see the explanation for the above comment in the readme, but why can; 't we call setUp from where ever we call `before_all`?

Konamiman replied:
If we call `setUp` here then it will be called again by PHPUnit itself later, and this could cause problems (or at least slow down the execution of the test suite).

vedanshujain replied:
ok, makes sense

vedanshujain commented on src/Testing/CodeHacking/Hacks/CodeHack.php:
```diff
@@ -0,0 +1,85 @@
+	 */
+	protected function tokenize( $code ) {
+		//phpcs:ignore PHPCompatibility.FunctionUse.NewFunctionParameters, PHPCompatibility.Constants.NewConstants
+		return PHP_VERSION_ID >= 70000 ? token_get_all( $code, TOKEN_PARSE ) : token_get_all( $code );
```
Since WC needs  minimum PHP 7, we can just raise an error if version is less than 7.0

Konamiman replied:
Ok, this code was in the original BypassFinals project.

vedanshujain replied:
feel free to remove it.

vedanshujain commented on src/Testing/CodeHacking/CodeHackerTestHook.php:
```diff
@@ -0,0 +1,167 @@
+final class CodeHackerTestHook implements BeforeTestHook, AfterTestHook {
+
+	public function executeAfterTest( string $test, float $time ): void {
+		if ( ! CodeHacker::has_persistent_hacks() ) {
```
We should clear nonpersistant hacks anyways right?

Konamiman replied:
Yes, and this happens in `executeBeforeTest`.

vedanshujain commented on src/Testing/CodeHacking/README.md:
```diff
@@ -0,0 +1,177 @@
+
+As you can see, the basic mechanism consists of creating a `public static before_[test_method_name]` method in the test class, where you register whatever hacks you need with `CodeHacker::add_hack` and finally you invoke `CodeHacker::enable()` to enable the hacking. `StaticMockerHack` and `FunctionsMockerHack` are two of the predefined hack classes inside the `Hack` folder, see their source code  ...
+
+You can define a `before_all` method too, which will run before all the tests (and before the `before_[test_method_name]` method).
```
let's call it something other than `before_all` if using setup is not possible,  because these two naming conventions are used interchangeably across languages.

Konamiman replied:
Ok, what would you suggest? Maybe `before_any_test`?

vedanshujain replied:
let's call them: `add_hack_for_[test_method_name]` and `add_hack_for_all`. I want to make them specific because for all normal use cases, `setup` and `teardown` should be used. wdyt?

vedanshujain reviewed (dismissed):
Awesome work @Konamiman ! Left some comments, I still have not tried it, will do during my second pass based on your feedback. Some general comments:
1. I have left comments for places that were not obvious to me, they generally start with "just curious". 
2. Maybe we should call this something else, like "mock" or "patch". Work hack has connotation that don't necessarily fit here. But maybe its too late now.
3. It could have been to good to reduce the scope of PR so that it can be easily reviewed.  For egs, we might not need to support annotations or static classes from the first go itself. But no worries.

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+<?php
```
Ok, maybe something like `tests/tools`, but I would keep namespaces.

vedanshujain replied:
sounds good :+1:

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+	 */
+	public static function add_hack( $hack, $persistent = false ) {
+		if ( ! is_callable( $hack ) && ! is_object( $hack ) ) {
+			throw new \Exception( "Hacks must be either functions, or objects having a 'process(\$text, \$path)' method." );
```
You mean adding `CodeHacker` in the error message?

vedanshujain replied:
yeah, but this is a nitpick, not a blocker. but it would be good to be consistent in all error messages.

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+	 * @param bool  $persistent If true, the hack will be registered as persistent (so that clear_hacks will not clear it).
+	 * @throws \Exception Invalid input.
+	 */
+	public static function add_hack( $hack, $persistent = false ) {
```
[I explain this in the README](https://github.com/woocommerce/woocommerce/blob/feature/code-hacker-for-unit-tests/src/Testing/CodeHacking/README.md#workaround-for-static-mocking-of-already-loaded-code-the-staticwrapper). The TLDR is that the code hacker works only for files that still haven't been loaded, but the tests bootstrap initializes Woo and this causes some files to be loaded, e.g. `WC_Admin_Settings`. As a workaround for the case of hacking static functions I have created a `StaticMockerHack` class that needs to be registered during bootstrap and be kept registered throughout the entire test suite run.

vedanshujain replied:
sorry, I am not super clear here. does this mean for static classes we need `$persistent` hacks?

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+		if ( $persistent ) {
+			self::$persistent_hacks[] = $hack;
+		}
+		self::$hacks[] = $hack;
```
The `$persistent_hacks` property is used only to keep track of which hacks are persistent, the `$hacks` property is the one actually used to apply the hacks. So persistent hacks need to be in both.

vedanshujain replied:
ah ok 👍

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+	 * @throws ReflectionException Error when instantiating ReflectionFunction.
+	 */
+	private static function is_valid_hack_callback( $callback ) {
+		return is_callable( $callback ) && 2 === ( new ReflectionFunction( $callback ) )->getNumberOfRequiredParameters();
```
Hack functions/methods take two parameters, one is the file contents and the other is the file path. Agreed on the constant.

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+		$ro = new ReflectionObject( ( $callback ) );
+		try {
+			$rm = $ro->getMethod( 'hack' );
+			return $rm->isPublic() && ! $rm->isStatic() && 2 === $rm->getNumberOfRequiredParameters();
```
I designed the hacker so that you register either functions or objects. Maybe accepting static methods would work too, but I wanted to keep the available options limited to a viable minimum.

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+			$rm = $ro->getMethod( 'hack' );
+			return $rm->isPublic() && ! $rm->isStatic() && 2 === $rm->getNumberOfRequiredParameters();
+		} catch ( ReflectionException $exception ) {
+			return false;
```
See `add_hack` method: when this returns false an exception with a proper error message is thrown. Also this is a private method and isn't used anywhere else.

vedanshujain replied:
[not a blocker] IMO it will be better to throw an error from as deep in the stack as possible and then bubble it up to the user (unless we need to mask the error which does not seem to be the case here).

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+	 * @param array $path_white_list Paths of the files to hack, can be relative paths.
+	 */
+	public static function set_white_list( array $path_white_list ) {
+		self::$path_white_list = $path_white_list;
```
This was present in [the original BypassFinals project](https://github.com/dg/bypass-finals) and I thought it wouldn't hurt to leave it in place.

vedanshujain replied:
let's remove it because it looks like a good way for someone to make an error and also we won't have to maintain it.

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+	 * @param string $path Specifies the URL that was passed to opendir().
+	 * @param int    $options Whether or not to enforce safe_mode (0x04).
+	 *
+	 * @return bool TRUE on success or FALSE on failure.
```
This method is part of [the stramWrapper "interface"](https://www.php.net/manual/en/class.streamwrapper.php). These methods need to be defined with the specified signature.

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+ */
+class CodeHacker {
+
+	const PROTOCOL = 'file';
```
I'm not sure but it looks like you can define a stream wrapper for any invented protocol e.g. `foobar` and specify it when opening the "file" by prepending the file name with `foobar://`.

vedanshujain replied:
Ah I see, I think it will good to add a comment here that we will overwrite the built-in `streamWrapper` for `file` protocol with class.

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+	 * @return false|string string representing the next filename, or FALSE if there is no next file.
+	 */
+	public function dir_readdir() {
+		return readdir( $this->handle );
```
For what I see `native` needs to be used only when there's a need to invoke a PHP function that has the same name as the method in the class. Keep in mind that this is a stream wrapper that for the most part is a "no-op", that is, most methods just redirect to the original PHP function (hence the "native" name, as in "redirect this to the native PHP functions").

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+	 *
+	 * @return mixed The underlying stream resource used by the wrapper, or FALSE.
+	 */
+	public function stream_cast( $cast_as ) {
```
Again, this method is defined in the streamWrapper "interface" with a fixed signature.

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+	 *
+	 * @return bool TRUE on success or FALSE on failure. If option is not implemented, FALSE should be returned.
+	 */
+	public function stream_metadata( $path, $option, $value ) {
```
Details here: https://www.php.net/manual/en/streamwrapper.stream-metadata.php

vedanshujain replied:
I see, is there any way to call the built-in `stream_metadata` here so that we don't have to maintain this across php versions?

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+	 */
+	public function stream_open( $path, $mode, $options, &$opened_path ) {
+		$use_path = (bool) ( $options & STREAM_USE_PATH );
+		if ( 'rb' === $mode && self::path_in_white_list( $path ) && 'php' === pathinfo( $path, PATHINFO_EXTENSION ) ) {
```
What do you mean by other modes? Directories perhaps?

vedanshujain replied:
I was thinking about protocols but based on other comments that do not make sense here because we will never use anything else than `file`. resolving this comment.

Konamiman commented on src/Testing/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,528 @@
+	 * @param int $arg1 Depends on $option.
+	 * @param int $arg2 Depends on $option.
+	 */
+	public function stream_set_option( $option, $arg1, $arg2 ) {
```
Again, the method must exist as it's defined in the streamWrapper "interface", even if it doesn't need any implementation. Looks like the original author thought that [the existing options](https://www.php.net/manual/en/streamwrapper.stream-set-option.php) weren't needed in a unit testing context.

Konamiman commented on src/Testing/CodeHacking/CodeHackerTestHook.php:
```diff
@@ -0,0 +1,167 @@
+		 */
+		$parts = explode( '::', $test );
+		if ( count( $parts ) < 2 ) {
+			return;
```
This is not an error. This check is in place because the hook is sometimes invoked with the word "Warning" instead of a class name, and these cases must be discarded.

vedanshujain replied:
> the hook is sometimes invoked with the word "Warning" instead of a class name
just out of curiosity, when will this happen?

Konamiman commented on src/Testing/CodeHacking/CodeHackerTestHook.php:
```diff
@@ -0,0 +1,167 @@
+		$class_name  = $parts[0];
+		$method_name = explode( ' ', $parts[1] )[0];
+
+		CodeHacker::clear_hacks();
```
The thing is that hacks can't be defined in `setUp` methods as by the time they run everything has been already loaded. That's why I added all the `before_` methods thing.

Konamiman commented:
> Maybe we should call this something else, like "mock" or "patch". Work hack has connotation that don't necessarily fit here. But maybe its too late now.

I chose a "scary" name on purpose so that this thing isn't used lightly (our priority is to write easily testable code and this is a last resort mechanism)

> It could have been to good to reduce the scope of PR so that it can be easily reviewed. For egs, we might not need to support annotations or static classes from the first go itself. But no worries.

I tried to divide the implementation in small(ish) commits to help with that, not sure if I succeeded though. 🙂

Konamiman committed:
Small improvements in the code hacker

- Convert a hardcoded number to a constant
- Add more information to an exception message
- Remove unnecessary PHPCS disabling
- Throw exception on PHP<7.0

Konamiman committed:
Move testing tools to the tests/Tools directory

The testing tools (only the code hacker at this time) have been moved
from 'src' to 'tests/Tools', since many opcode cache plugins
load the whole src folder in production.

Also, an extra autoloader is set in the tests bootstrap so that
the 'tests/Tools' directory corresponds, using PSR4, to the
'Automattic\WooCommerce\Testing\Tools' namespace.

vedanshujain commented on tests/Tools/CodeHacking/CodeHacker.php:
```diff
@@ -0,0 +1,529 @@
+ * @package WooCommerce/Testing
+ */
+
+//phpcs:disable WordPress.WP.AlternativeFunctions, WordPress.PHP.NoSilencedErrors.Discouraged
```
let's not disable phpcs for entire hacks, if there is good reason to do so we should mention it.

claudiosanches replied:
I prefer to ignore those in each line, but since this is an internal tool, sounds fine to remove in all instances.
But note that doing it in our code base isn't recommended, since we can use PHP CodeSniffer to detect those instances and help us write better code.

Konamiman committed:
Code hacker resdesign for single-load of code files.

The code hacker as originally designed, as a mechanism that allowed
to enable hacks at the individual test level, is flawed because it
assumes that code files are loaded before each test, but actually
the PHP engine loads code files only once.

Therefore this commit redesigns it so that the two existing main hacks,
the functions mocker and the static methods hacker, are applied
to all the relevant functions and classes at bootstrap time, and
mocks for each individual function/method can be registered at the
beginning of each test. See README for the full details.

claudiosanches commented:
@Konamiman great work! But seems like this code is causing PHP fatal errors on Travis. After that I think I can finish review and merge this PR. Thanks!

claudiosanches reviewed (approved):
Great work!

There's only one error on Travis, but related to e2e tests, see:
> TimeoutError: Element #message (text: "1 coupon moved to the Trash.") not found

Sounds like nothing that we need to worry for this PR.

Konamiman committed:
Update autoload-dev in composer.json to add tests/tools directory

The mapping of the "Automattic\WooCommerce\Testing\Tools\" namespace
to the "tests/tools" directory is moved from manual registration
inside the tests bootstrap constructor to a declaration inside the
autload-dev section in composer.json.
